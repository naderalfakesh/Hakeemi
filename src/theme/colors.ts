export const palette = {
  primary: {
    base: 'rgba(0, 188, 212, 1)',
    light: 'rgba(214, 250, 255, 1)',
    dark: 'rgba(144, 180, 206, 1)',
  },
  secondary: {
    base: 'rgba(237, 243, 247, 1)',
  },
  danger: {
    base: 'rgba(255, 87, 34, 1)',
  },
  grey: {
    base: 'rgba(240, 240, 240, 1)',
    light: 'rgba(251, 251, 251, 1)',
    dark: 'rgba(144, 144, 144, 1)',
  },
  white: {
    base: 'rgba(255, 255, 255, 1)',
  },
  black: {
    base: 'rgba(0, 0, 0, 1)',
  },
};

export default {
  primary: palette.primary.base,
  secondary: palette.secondary.base,
  danger: palette.danger.base,
  light: palette.grey.light,
  dark: palette.grey.dark,
  white: palette.white.base,
  black: palette.black.base,
};
