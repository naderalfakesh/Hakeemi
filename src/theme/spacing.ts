const spacing = {
  /** Space 2 */
  tiny: 2,
  /** Space 4 */
  mini1: 4,
  /** Space 6 */
  mini2: 6,
  /** Space 8 */
  mini3: 8,
  /** Space 10 */
  small1: 10,
  /** Space 12 */
  small2: 12,
  /** Space 16 */
  small3: 16,
  /** Space 18 */
  small4: 18,
  /** Space 20 */
  medium1: 20,
  /** Space 24 */
  medium2: 24,
  /** Space 32 */
  medium3: 32,
  /** Space 40 */
  big1: 40,
  /** Space 50 */
  big2: 50,
  /** Space 68 */
  big3: 68,
  /** Space 76 */
  huge1: 76,
  /** Space 88 */
  huge2: 88,
  /** Space 92 */
  huge3: 92,
};
export default spacing;
