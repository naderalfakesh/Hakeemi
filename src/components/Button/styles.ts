import { StyleSheet, ViewStyle } from 'react-native';
import colors from '../../theme/colors';

export type Sizes = 'small' | 'medium' | 'big';
export type Themes = 'primary' | 'secondary' | 'grey';

export default StyleSheet.create({
  base: {
    height: 72,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 20,
  },
});

export const themeStyle: Record<Themes, ViewStyle> = StyleSheet.create({
  primary: {
    backgroundColor: colors.primary,
  },
  secondary: {
    backgroundColor: colors.white,
  },
  grey: {
    backgroundColor: colors.light,
  },
});

export const pressedStyle: Record<Themes, ViewStyle> = StyleSheet.create({
  primary: {
    backgroundColor: colors.light,
  },
  secondary: {
    backgroundColor: colors.secondary,
  },
  grey: {
    backgroundColor: colors.dark,
  },
});
