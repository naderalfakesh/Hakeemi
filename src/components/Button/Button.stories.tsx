import React, { Fragment } from 'react';
import { View, StyleSheet } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import Button from './Button';
import { Themes } from './styles';

const Container = (storyFn: any) => (
  <View style={styles.container}>{storyFn()}</View>
);

const ButtonDemo = ({ theme }: { theme: Themes }) => (
  <Fragment>
    <Button theme={theme} size="small">
      Small button
    </Button>
    <Button theme={theme} size="medium">
      Medium Button
    </Button>
    <Button theme={theme} size="big">
      Big button
    </Button>
    <View style={styles.icons}>
      <Button theme={theme} size="squareIcon" iconName="heart-beat">
        Heartbeats
      </Button>
      <Button theme={theme} size="inlineIcon" iconName="heart-beat">
        Heartbeats
      </Button>
    </View>
  </Fragment>
);

storiesOf('Buttons', module)
  .addDecorator(Container)
  .add('primary', () => <ButtonDemo theme="primary" />)
  .add('secondary', () => <ButtonDemo theme="secondary" />)
  .add('grey', () => <ButtonDemo theme="grey" />);

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#e5e5e5',
    flexGrow: 1,
    padding: 8,
    justifyContent: 'space-evenly',
  },
  icons: {
    flexDirection: 'row',
    justifyContent: 'space-around',
  },
});
