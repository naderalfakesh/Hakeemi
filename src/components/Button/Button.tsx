import React, { FC } from 'react';
import {
  Pressable,
  ViewStyle,
  PressableProps,
  ActivityIndicator,
} from 'react-native';
import Icon, { IconName } from '../Icon';
import Text, { Props as TextProps } from '../Text';
import styles, {
  pressedStyle,
  Sizes,
  Themes,
  themeStyle,
  sizeStyle,
  iconColor,
} from './styles';

interface Props extends PressableProps {
  style?: ViewStyle;
  theme?: Themes;
  size?: Sizes;
  iconName?: IconName;
  highlight?: boolean;
  loading?: boolean;
  onPress?: () => void;
  subtitle?: string;
}

const mapTextColor = (theme: Themes): TextProps['color'] => {
  switch (theme) {
    case 'primary':
      return 'white';
    case 'secondary':
      return 'secondary';
    case 'tertiary':
      return 'primary';
    case 'grey':
      return 'dark';
  }
};

type sizeProps = {
  size: TextProps['size'];
  level: TextProps['level'];
};

const mapTextSize = (size: Sizes): sizeProps => {
  switch (size) {
    case 'small':
      return { size: 'body-16', level: '600' };
    case 'medium':
      return { size: 'body-16', level: '600' };
    case 'big':
      return { size: 'button-22', level: '600' };
    case 'squareIcon':
      return { size: 'button-10', level: '400' };
    case 'inlineIcon':
      return { size: 'body-16', level: '400' };
  }
};

const mapSubtitleSize = (size: Sizes): sizeProps => {
  switch (size) {
    case 'small':
      return { size: 'body-14', level: '400' };
    case 'medium':
      return { size: 'body-14', level: '400' };
    case 'big':
      return { size: 'button-8', level: '400' };
    case 'squareIcon':
      return { size: 'button-10', level: '400' };
    case 'inlineIcon':
      return { size: 'body-16', level: '400' };
  }
};

const Button: FC<Props> = ({
  children,
  style,
  size = 'medium',
  theme = 'primary',
  iconName,
  onPress,
  subtitle,
  highlight = true,
  loading = false,
  ...otherProps
}) => {
  return (
    <Pressable
      testID="Button"
      style={({ pressed }) => [
        styles.base,
        themeStyle[theme],
        sizeStyle[size],
        !loading && pressed && highlight && pressedStyle[theme],
        size === 'inlineIcon' && styles.inlineIcon,
        style,
      ]}
      onPress={!loading && onPress ? onPress : undefined}
      {...otherProps}>
      {iconName && (
        <Icon
          style={styles.icon}
          name={iconName}
          size={24}
          color={iconColor[theme].color}
        />
      )}
      {loading ? (
        <ActivityIndicator size="large" color="white" />
      ) : (
        <Text color={mapTextColor(theme)} {...mapTextSize(size)}>
          {children}
        </Text>
      )}
      {subtitle && (
        <Text color={mapTextColor(theme)} {...mapSubtitleSize(size)}>
          {subtitle}
        </Text>
      )}
    </Pressable>
  );
};

export default Button;
