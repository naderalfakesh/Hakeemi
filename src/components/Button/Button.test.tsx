import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Button from './Button';
import colors from '../../theme/colors';
import spacing from '../../theme/spacing';
import { StyleSheet } from 'react-native';

describe('Button component test', () => {
  it('renders correctly with default params', () => {
    const { toJSON } = render(<Button>Hello</Button>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders primary small button', () => {
    const { getByTestId } = render(
      <Button theme="primary" size="small">
        Hello
      </Button>,
    );
    const btn = getByTestId('Button');
    const style = StyleSheet.flatten(btn.props.style);
    expect(style.backgroundColor).toBe(colors.primary);
    expect(style.padding).toBe(spacing.small1);
  });

  it('renders secondary medium small button', () => {
    const { getByTestId } = render(
      <Button theme="secondary" size="medium">
        Hello
      </Button>,
    );
    const btn = getByTestId('Button');
    const style = StyleSheet.flatten(btn.props.style);
    expect(style.backgroundColor).toBe(colors.white);
    expect(style.padding).toBe(spacing.small3);
  });

  it('renders grey big small button', () => {
    const { getByTestId } = render(
      <Button theme="grey" size="big">
        Hello
      </Button>,
    );
    const btn = getByTestId('Button');
    const style = StyleSheet.flatten(btn.props.style);
    expect(style.backgroundColor).toBe(colors.light);
    expect(style.padding).toBe(spacing.medium1);
  });

  it('overrides styles', () => {
    const { getByTestId } = render(
      <Button style={{ borderRadius: 0 }}>Hello</Button>,
    );
    const btn = getByTestId('Button');
    const style = StyleSheet.flatten(btn.props.style);
    expect(style.borderRadius).toBe(0);
  });

  it('invokes a function when pressed', () => {
    const mockFn = jest.fn();
    const { getByText } = render(<Button onPress={mockFn}>BUTTON</Button>);
    const btn = getByText('BUTTON');
    fireEvent.press(btn);
    expect(mockFn).toBeCalled();
  });
});
