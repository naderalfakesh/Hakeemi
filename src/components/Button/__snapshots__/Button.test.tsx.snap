// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test renders correctly with default params 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 16,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#00bbd4",
      },
      Object {
        "padding": 16,
      },
      false,
      false,
      undefined,
    ]
  }
  testID="Button"
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "System",
        "fontSize": 16,
        "fontWeight": "600",
        "includeFontPadding": false,
        "lineHeight": 20,
        "textAlign": undefined,
      }
    }
    testID="Text"
  >
    Hello
  </Text>
</View>
`;
